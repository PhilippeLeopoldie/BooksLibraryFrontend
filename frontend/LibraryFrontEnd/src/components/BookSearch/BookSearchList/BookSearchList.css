.bookSearchListContainer--black {
    color: #ffffff;
    background-color: inherit;
}

.bookSearchListContainer--black > h2 {
    color: #ffffff;
    background-color: inherit;
}


.bookSearchListContainer {
    display: flex;
    overflow-x: auto;
    -webkit-overflow-scrolling: touch;
    background-color: inherit;
}

.bookSearchListContainer--center {
    display: flex;
    justify-content: center;
    background-color: inherit;
}

@media screen and (min-width: 700px) {

    .bookSearchListContainer {
        grid-template-columns: repeat(2, auto);
        margin-bottom: 0%;
    }
}

@media screen and (min-width: 1024px) {

    .bookSearchListContainer {
        display: grid;
        justify-content: center;
    }

    .bookSearchListContainer {
        grid-template-columns: repeat(3, auto);
    }
}

@media screen and (min-width: 1300px) {
    .bookSearchListContainer {
        grid-template-columns: repeat(4, auto);
    }
}

@media screen and (min-width: 1588px) {
    .bookSearchListContainer {
        grid-template-columns: repeat(5, auto);
    }
}
