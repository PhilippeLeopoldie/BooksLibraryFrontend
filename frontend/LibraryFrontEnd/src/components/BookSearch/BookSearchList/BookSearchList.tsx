import { BookCard } from "../../BookCard/BookCard";
import "../../BookSearch/BookSearchList/BookSearchList.css";
import { useContext, useEffect, useState } from "react";
import { BOOK_BY_TITLE_OR_AUTHOR_URL } from "../../../constants/api";
import { ThemeContext } from "../../../App/App";

type BooksType = {
    book: {
        id: number;
        title: string;
        author: string;
        averageRate: number;
        description: string;
    };
};

type BooksSearchCriteria = {
    titleOrAuthor: string;
};

export const BookSearchList = ({ titleOrAuthor }: BooksSearchCriteria) => {

    const [books, setBooks] = useState<BooksType[] | null>([]);
    const searching = "Searching ...";
    const [waitingMessage, setWaitingMessage] = useState<string>(searching);
    const theme = useContext(ThemeContext);
    const fetchBooksByTitleOrAuthor = async () => {
        try {
            if (titleOrAuthor === "") {
                setBooks([]);
                setWaitingMessage(searching);
            }
            if (titleOrAuthor !== "") {
                const booksResponse: Response = await fetch(
                    BOOK_BY_TITLE_OR_AUTHOR_URL + titleOrAuthor
                );
                if (booksResponse.status === 200) {
                    const booksResponseData = await booksResponse.json();
                    setBooks(booksResponseData);
                    setWaitingMessage("");
                } else if (booksResponse.status === 404) {
                    setBooks([]);
                    setWaitingMessage("");
                    console.log(booksResponse);
                }
            }
        } catch (error) {
            console.error("Error fetching books:", error);
        }
    };

    useEffect(() => {
        setWaitingMessage(searching);
        const debounceTimer = setTimeout(() => {
            fetchBooksByTitleOrAuthor();
        }, 2000);
        return () => clearTimeout(debounceTimer);
    }, [titleOrAuthor]);

    if (titleOrAuthor === "") {
        return <h1 className={"bookSearchListContainer--" + theme}></h1>;
    }
    if (titleOrAuthor != "" && waitingMessage) {
        return <h2 className={"bookSearchListContainer--" + theme}>{waitingMessage}</h2>;
    }

    return (
        <>
            <div className={"bookSearchListContainer--" + theme}>
                <h2>
                    {books && books.length > 1
                        ? `(${books.length}) books found`
                        : `(${books?.length}) book found`}
                </h2>
                <div className={books && books.length > 1 ? "bookSearchListContainer" : "bookSearchListContainer--center"}>
                    {books &&
                        Array.isArray(books) &&
                        books
                            .sort((a, b) => b.book.id - a.book.id)
                            .map((bookDetail) => (
                                <BookCard
                                    key={bookDetail.book.id}
                                    book={bookDetail.book}
                                />
                            ))}
                </div>
            </div>
        </>
    );
};
