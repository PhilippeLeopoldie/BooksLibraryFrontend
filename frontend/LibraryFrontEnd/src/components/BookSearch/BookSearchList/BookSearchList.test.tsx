import {render, screen,} from "@testing-library/react";
import { BookSearchList } from "./BookSearchList";
import "@testing-library/jest-dom";
import { useState } from "react";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

type BooksType = {
  book: {
    id: number;
    title: string;
    author: string;
    averageRate: number;
  };
};

const book : BooksType = {
  book:{
    id: 1,
    title: "title",
    author: "author",
    averageRate: 3,
  }

}

describe("BookListSearch", () => {
  test("snapShot: render Loading... while fetching", ()=> {
    (useState as jest.Mock).mockImplementation(() => [book, jest.fn()]);
    (useState as jest.Mock).mockImplementation(() => ["Loading...", jest.fn()]);

    const {asFragment} = render(<BookSearchList titleOrAuthor={"title"}/>)

    expect(asFragment()).toMatchSnapshot();
  })

  test("snapShot: should not render Loading... when not fetching", ()=> {
    (useState as jest.Mock).mockImplementation(() => [[], jest.fn()]);
    (useState as jest.Mock).mockImplementation(() => ["Loading...", jest.fn()]);

    const {asFragment} = render(<BookSearchList titleOrAuthor={""}/>)

    expect(asFragment()).toMatchSnapshot();
  })

  test("snapShot: render (0) book found", ()=> {
    (useState as jest.Mock).mockImplementation(() => [[], jest.fn()]);
    (useState as jest.Mock).mockImplementation(() => ["", jest.fn()]);

    const {asFragment} = render(<BookSearchList titleOrAuthor={"z"}/>)

    expect(asFragment()).toMatchSnapshot();
  })

  test("snapShot: render (1) book found", ()=> {
    (useState as jest.Mock).mockImplementation(() => [book, jest.fn()]);
    (useState as jest.Mock).mockImplementation(() => ["", jest.fn()]);

    const {asFragment} = render(<BookSearchList titleOrAuthor={"chat"}/>)

    expect(asFragment()).toMatchSnapshot();
  })

  
})