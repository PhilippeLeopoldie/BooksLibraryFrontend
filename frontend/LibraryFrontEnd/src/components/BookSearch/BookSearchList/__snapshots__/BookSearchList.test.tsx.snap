// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookListSearch snapShot: render (0) book found 1`] = `
<DocumentFragment>
  <div
    class="bookListSearchContainer--natural"
  >
    <h2>
      (0) book found
    </h2>
    <div
      class="bookListContainer-center"
    />
  </div>
</DocumentFragment>
`;

exports[`BookListSearch snapShot: render (1) book found 1`] = `
<DocumentFragment>
  <div
    class="bookListSearchContainer--natural"
  >
    <h2>
      (0) book found
    </h2>
    <div
      class="bookListContainer-center"
    />
  </div>
</DocumentFragment>
`;

exports[`BookListSearch snapShot: render Loading... while fetching 1`] = `
<DocumentFragment>
  <h1
    class="bookListSearchContainer--natural"
  >
    Loading...
  </h1>
</DocumentFragment>
`;

exports[`BookListSearch snapShot: should not render Loading... when not fetching 1`] = `
<DocumentFragment>
  <h1
    class="bookListSearchContainer--natural"
  />
</DocumentFragment>
`;
